taichi/backends/cuda/cuda_profiler.cpp

```diff
@@ -1,19 +1,167 @@
 #include "taichi/backends/cuda/cuda_profiler.h"
 #include "taichi/backends/cuda/cuda_driver.h"
 #include "taichi/backends/cuda/cuda_context.h"
-#include "taichi/system/timeline.h"
 
 TLANG_NAMESPACE_BEGIN
 
 #if defined(TI_WITH_CUDA)
 
-std::string KernelProfilerCUDA::title() const {
-  return "CUDA Profiler";
+bool check_device_capability() {
+  void *device;
+  int cc_major;
+  CUDADriver::get_instance().device_get(&device, 0);
+  CUDADriver::get_instance().device_get_attribute(
+      &cc_major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, device);
+  if (cc_major < 7) {
+    TI_WARN(
+        "CUPTI profiler APIs unsupported on Device with compute capability < "
+        "7.0 , fallback to default kernel profiler");
+    return false;
+  }
+  return true;
+}
+
+KernelProfilerCUDA::KernelProfilerCUDA() {
+// if Taichi was compiled with CUDA toolit, then use CUPTI
+// TODO : add set_mode() to select toolkit by user
+#if defined(TI_WITH_CUDA_TOOLKIT)
```

Comment:
Btw with this logic here, is cudaEvent still the default profiling method (as described in the comment)? Looks like if I install cuda toolkit, my default is cupti?

Reply:
one more prerequisite in needed ( compile Taichi with ):
`TAICHI_CMAKE_ARGS="-DTI_WITH_CUDA_TOOLKIT:BOOL=ON" python3 setup.py develop --user`

Reply:
some fallback logic will be added in the next PR : )

Reply:
`TI_WITH_CUDA_TOOLKIT` now means `TI_COMPILE_WITH_CUDA_TOOLKIT`
I'll refactor the logic here after migrate to `FindCUDA.cmake`  #2899 ( not done yet )

Reply:
SG!